src/main.rs

```diff
@@ -70,6 +86,31 @@ fn write_state(infile: &str, outfile: &str) {
     let mut min_closeness: f64 = 10000.0;
     let mut max_closeness: f64 = 0.0;
 
+    let res = geo_cache.load().await;
+    if res.is_err() {
+        println!("No cache file to load! Will be created one.");
+    }
+
+    //TODO(asmie): crunchy should be more configurable - currently IT is needed to have IP2LOCATION-LITE-DB11.BIN
+    // in the current directory. As IPS will have many configurable factors it should be possible to
+    // set them easily from configuration file.
+    geo_cache.add_provider(Box::new(Ip2LocationService::new(
+        "IP2LOCATION-LITE-DB11.BIN",
+    )));
+
+    //TODO(asmie): enabling providers below can cause program to wait extremely long time for response
+    // when exceeding free rate limit. Currently use it only when don't have IP2LOCATION-LITE-DB11.BIN.
+    // That's why they are added on the further places. In the future there should be some timeout mechanism
+    // for each IP address to avoid waiting too long time for response.
```

Comment:
some long-term ideas:
we could create a fully independent IP info service (cloud IP cache server) which would be running 24/7 on some cloud server

our services - like this one here would ask it via REST API queries for IP address info
the 'cloud cache server' would maintain a database which would be updated periodically from the online providers for the IP addresses it was queried previously - which we can feed for the first time with different crawlers

Reply:
That's generally good idea which could allow slow cache updates respecting limits of particular services. I like it.

However, it depends on the project's future. If it'd be used broadly in our projects, well, there'd be a good reason to do this. If we do not plan increasing accuracy or using it somewhere else I'd put it into TODO list.

Reply:
> TODO list.

yeah, this is a project, not a task. let's create another view in our improvement bucket these ideas - so we don't forget about those

Reply:
Agree.